File: leibnz/sequencial_pi.m
% Qualquer arquivo .m que começa com uma definição de função é um arquivo de função.
% Adicionar uma declaração sem sentido ao topo transforma em um script.
0;

function [aprox_pi, duration] = pi_aprox_sequencial(n)
  start = tic;

  soma = 0;

  for i = 0:n
    soma = soma + ((-1)^i / (2*i + 1));
  end

  aprox_pi = 4 * soma;

  duration = toc(start);
end


%
% Código de teste
disp("\n--Versão iterativa--")
faixa = [20, 50, 100, 1000, 25000, 50000, 100000, 500000, 1000000, 10000000];
times = zeros(size(faixa));

for i = 1:length(faixa)
  n = faixa(i);
  [pi_aprox, time_duration] = pi_aprox_sequencial(n);
  times(i) = time_duration;

  disp(["## it ", num2str(i), ": ", "n=", num2str(n)])
  disp(["Aproximacao: ", num2str(pi_aprox)])
  disp(["Tempo de execucao: ", num2str(time_duration)])
  disp(" ");
end

plot(faixa, times);
xlabel('n');
ylabel('Tempo (s)');
title('Tempos de execução em relação a n');

File: montecarlo/sequencial_pi.py
import random
import time
import math
import matplotlib.pyplot as plt

if __name__ == '__main__':
  faixa = [100000, 1000000, 10000000, 100000000]
  times = []
  
  for i in range(len(faixa)):
    start = time.time()

    n = faixa[i]
    acertos = 0

    for _ in range(n):
      x = random.random()
      y = random.random()

      if (x*x + y*y < 1):
        acertos+=1

    aprox_pi = 4 * acertos / n
    erro = math.pi - aprox_pi

    end = time.time()

    times.append(end - start)   

    print('-it %d:  n=%d' % (i, n))
    print("Aproximacao: ", aprox_pi)
    print("Tempo de execucao: ", end - start, "\n")

  plt.plot(faixa, times)
  plt.xlabel('n')
  plt.ylabel('Tempo de execução (s)')
  plt.title('Tempo de execução em função do número de pontos')
  plt.show()
